### PR TITLE
Eng 19016 topic key columns

### PR DESCRIPTION
**Pull request by rdykiel**
Support for topic key-columns

### CODE REVIEW

File: src/frontend/org/voltdb/export/ExportAvroSerializer.java [hunk 1]
```diff
@@ -30,14 +36,9 @@
...
 
-import java.io.ByteArrayOutputStream;
-import java.io.IOException;
-import java.nio.ByteBuffer;
-import java.util.Arrays;
-import java.util.List;
-import java.util.Map;
-import java.util.Objects;
-import java.util.concurrent.ConcurrentHashMap;
+import io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient;
```
Comment (n3world) [inline src/frontend/org/voltdb/export/ExportAvroSerializer.java]: What is this sort order for imports. It isn't alphabetical or io would be before org?

File: src/frontend/org/voltdb/exportclient/ExportRow.java [hunk 2]
```diff
@@ -107,6 +110,89 @@ public Long getTimestamp() {
...
+     * in whatever target namespace the row will be decoded into.
+     *
+     * @param keyColumnNames list of key column names in the desired order for the key
+     * @return the {@link ExportRow} instance carrying the data for the key columns
+     */
+    public ExportRow extractKey(List<String> keyColumnNames) {
+        String keyName = tableName + "_key";
+        List<String> colNames = new ArrayList<>();
+        List<Integer> colLengths = new ArrayList<>();
+        List<VoltType> colTypes = new ArrayList<>();
```
Comment (n3world) [inline src/frontend/org/voltdb/exportclient/ExportRow.java]: Not your fault but this is really awkward maintaining three lists that have to be kept in sync

File: src/frontend/org/voltdb/exportclient/ExportRow.java [hunk 3]
```diff
@@ -107,6 +110,89 @@ public Long getTimestamp() {
...
+        }
+        else if (partitionValue != null) {
+            // Use the partition column value
+            colNames.add(names.get(partitionColIndex));
+            colLengths.add(lengths.get(partitionColIndex));
+            colTypes.add(types.get(partitionColIndex));
+            colValues[INTERNAL_FIELD_COUNT] = partitionValue;
+        }
+        else {
+            // Use a fake column with the partition
```
Comment (n3world) [inline src/frontend/org/voltdb/exportclient/ExportRow.java]: Why is this making a fake key? Why not just have a null key in this case

File: src/frontend/org/voltdb/exportclient/ExportRow.java [hunk 4]
```diff
@@ -107,6 +110,89 @@ public Long getTimestamp() {
...
+            colLengths.add(lengths.get(i));
+            colTypes.add(types.get(i));
+            colValues[i] = values[i];
+        }
+
+        // Build key data
+        if (hasCols) {
+            // Use the listed key columns, in the order specified
+            Map<String, Integer> colMap = new HashMap<>();
+            for (int i = 0; i < names.size(); i++) {
```
Comment (n3world) [inline src/frontend/org/voltdb/exportclient/ExportRow.java]: Minor: i++ is slightly more expensive than ++i

File: src/frontend/org/voltdb/export/ExportAvroSerializer.java (hunk 1 above)
Comment (rdykiel) [reply src/frontend/org/voltdb/export/ExportAvroSerializer.java]: Eclipse order import left them unchanged so I figure there must be some underlying convention. Will leave as is

File: src/frontend/org/voltdb/exportclient/ExportRow.java (hunk 3 above)
Comment (rdykiel) [reply src/frontend/org/voltdb/exportclient/ExportRow.java]: Preserving previous behavior. When no partition data exists we use the value of the partition ID.

File: src/frontend/org/voltdb/exportclient/ExportRow.java (hunk 4 above)
Comment (rdykiel) [reply src/frontend/org/voltdb/exportclient/ExportRow.java]: Now we're really splitting hair

File: src/frontend/org/voltdb/exportclient/ExportRow.java (hunk 2 above)
Comment (rdykiel) [reply src/frontend/org/voltdb/exportclient/ExportRow.java]: The current structure is expected by all the decoders so not changing this right now.

File: src/frontend/org/voltdb/exportclient/ExportRow.java (hunk 3 above)
Comment (n3world) [reply src/frontend/org/voltdb/exportclient/ExportRow.java]: Wouldn't no partition data mean it is replicated and not partitioned and then the partition ID doesn't really mean anything

**Pull request merged**